### ISSUE TITLE
MP REST Client and mocking

### ISSUE THREAD

**Issue by FroMage** [quarkusio/quarkus#17861]
The guide at https://quarkus.io/guides/rest-client says:

```java
public class WiremockCountries implements QuarkusTestResourceLifecycleManager {  

	private WireMockServer wireMockServer;

	@Override
	public Map<String, String> start() {
		wireMockServer = new WireMockServer();
		wireMockServer.start(); 

		stubFor(get(urlEqualTo("/v2/name/GR"))   
				.willReturn(aResponse()
						.withHeader("Content-Type", "application/json")
						.withBody(
                        "[{" +
                        "\"name\": \"Ελλάδα\"," +
                        "\"capital\": \"Αθήνα\"" +
                        "}]"
						)));

		stubFor(get(urlMatching(".*")).atPriority(10).willReturn(aResponse().proxiedFrom("https://restcountries.eu/rest")));   

		return Collections.singletonMap("org.acme.rest.client.CountriesService/mp-rest/url", wireMockServer.baseUrl()); 
	}

	@Override
	public void stop() {
		if (null != wireMockServer) {
			wireMockServer.stop();  
		}
	}
}
```

This is a nice feature, but I would expect we improve our UX here and allow users to do some much simpler mock set-up, such as:

```java
public class WiremockCountries {  

    @RestMock
    public void mockCountries(CountriesService service) {
        Country greece = new Country();
        greece.name = "Ελλάδα";
        greece.capital = "Αθήνα";
        Set<Country> set = new HashSet<>();
        set.add(greece);
        RestMocker.when(service.getByName("GR")).replyWith(set);
        RestMocker.forward().toOriginalHost();
    }
}
```

So, we'd collect all `@RestMock` methods, see what service they mock, provide them with an empty stub for the service, then implement a Mockito-like `RestMocker` class allowing us to set up HTTP mocks, which ultimately forwards to `WireMockServer`, but allows us to mock several remote services in one go, and not muck about with JSON in strings.

WDYT?

/CC @geoand @stuartwdouglas @michalszynkiewicz

**Comment by geoand** [comment]
Absolutely +!

**Comment by geoand** [comment]
I can start working on this next week

**Comment by geoand** [comment]
I assume this would be a new `quarkus-rest-client-test` module?

**Comment by FroMage** [comment]
I imagined it would go over the wire, because this is an HTTP proxy, after all. Here we're setting up fake HTTP responses for anything that connects to that URL we're setting up. You could turn the REST client itself into a mock, but we can already do that ATM with regular mocks, no?

**Comment by geoand** [comment]
Going over the wire would obviously be the best type of test. 

So I assume what we want is to setup wiremock to return the JSON representation of the object we provide, right?

**Comment by FroMage** [comment]
Yes.

**Comment by geoand** [comment]
Gotcha. I'll hopefully start looking at it tomorrow

**Comment by geoand** [comment]
I think `RestMocker.forward().toOriginalHost();` needs to be `RestMocker.forward(countriesService).toOriginalHost();`, right?

**Comment by geoand** [comment]
I am also wondering if this should only be for the reactive-rest-client or for both.

The advantage of doing it only for the former is that we can tie into all the processed metadata and avoid having to re-parse annotations and such.

WDYT @FroMage @stuartwdouglas @michalszynkiewicz ?

**Comment by FroMage** [comment]
Good question. Probably start with the reactive one and later expand if we need to? I assume we'll be able to reuse all the code, we'd just have to add new code for the parsing?

**Comment by geoand** [comment]
From the quick pass I did, for the old rest client we don't really keep metada around for the methods as we do for the new one.

**Comment by geoand** [comment]
I'll start working on this tomorrow if nothing else urgent comes up

**Comment by michalszynkiewicz** [comment]
I stayed quiet because I'm struggling to see what's the main use case for this.

- proxying a real service - it seems to be a good match for it but I have never seen it in practice, so it seems to be a niche for me. But I may be wrong, of course
- what I think people usually test with wiremock is if the annotations they put on the client interface work properly. When we generate a service from this set of annotations, this doesn't work
- with testing what reaches a rest client, wouldn't someone just mock the client?

**Comment by FroMage** [comment]
> proxying a real service - it seems to be a good match for it but I have never seen it in practice, so it seems to be a niche for me. But I may be wrong, of course

Not a niche at all, in fact I've seen it in k8s operator land where they're really mocking a lot of remote stuff. In fact if you go micro-services you just absolutely need to mock a lot of HTTP calls.

> what I think people usually test with wiremock is if the annotations they put on the client interface work properly. When we generate a service from this set of annotations, this doesn't work

Well, yes it would, because we're only using the mock to set up the HTTP mocking, not to make any calls. All the calls will be via the normal clients and go through HTTP to wiremock

> with testing what reaches a rest client, wouldn't someone just mock the client?

I did wonder that myself, TBH. And I'm not 100% of the answer. I guess part of the answer is that if you use Mockito on the client interface you will not be able to mock stuff that isn't in the return type of the method?

Like, for example, suppose you want to mock HTTP headers or cookies being returned by the server, you can't do that by mocking the client interface, because you can only mock the return type, which is likely an entity. So if you have client filters they will never see the headers. Actually I'm not even sure if the clients filters get called if you mock the client interface.

So this is a bit more powerful? It's certainly different. It also allows you to easily define HTTP mocks even if the tested code isn't using MP-REST Client, such as lower level HTTP libs or custom clients, such as a k8s client? In this case you'd probably generate the MP-REST Client interface only for your tests, but :shrug:.

**Comment by michalszynkiewicz** [comment]
> > proxying a real service - it seems to be a good match for it but I have never seen it in practice, so it seems to be a niche for me. But I may be wrong, of course
> 
> Not a niche at all, in fact I've seen it in k8s operator land where they're really mocking a lot of remote stuff. In fact if you go micro-services you just absolutely need to mock a lot of HTTP calls.

mock, sure, but proxy?

> 
> > what I think people usually test with wiremock is if the annotations they put on the client interface work properly. When we generate a service from this set of annotations, this doesn't work
> 
> Well, yes it would, because we're only using the mock to set up the HTTP mocking, not to make any calls. All the calls will be via the normal clients and go through HTTP to wiremock

Isn't the proposal using the client interface for setting up wiremock?


> 
> > with testing what reaches a rest client, wouldn't someone just mock the client?
> 
> I did wonder that myself, TBH. And I'm not 100% of the answer. I guess part of the answer is that if you use Mockito on the client interface you will not be able to mock stuff that isn't in the return type of the method?
> 
> Like, for example, suppose you want to mock HTTP headers or cookies being returned by the server, you can't do that by mocking the client interface, because you can only mock the return type, which is likely an entity. So if you have client filters they will never see the headers. Actually I'm not even sure if the clients filters get called if you mock the client interface.
> 
> So this is a bit more powerful? It's certainly different. It also allows you to easily define HTTP mocks even if the tested code isn't using MP-REST Client, such as lower level HTTP libs or custom clients, such as a k8s client? In this case you'd probably generate the MP-REST Client interface only for your tests, but .

We're getting really close to creating a JAX-RS resource class in the test here, which is already easily doable ;)

**Comment by geoand** [comment]
> We're getting really close to creating a JAX-RS resource class in the test here, which is already easily doable ;)

Even if that is the case, why not give users a nice DSL in order to make it super easy? 
Much as we made mocking much easier.

Furthermore, with something like this, we can even think of injecting network faults, errorneous responses, delays etc... that you can't do with regular mocks (at least not at the right layer) and would require users to setup a lot of extra code if they did in with a custom JAX-RS Resource as you mention

**Comment by michalszynkiewicz** [comment]
> > We're getting really close to creating a JAX-RS resource class in the test here, which is already easily doable ;)
> 
> Even if that is the case, why not give users a nice DSL in order to make it super easy?
> Much as we made mocking much easier.
> 
> Furthermore, with something like this, we can even think of injecting network faults, errorneous responses, delays etc... that you can't do with regular mocks (at least not at the right layer) and would require users to setup a lot of extra code if they did in with a custom JAX-RS Resource as you mention

That's a good point. It's rather for testing if Quarkus behaves as it should, I think. But yes, we can simplify stuff a lot here.


Would be great if we could also do something for easy testing for WebSockets :)

**Comment by michalszynkiewicz** [comment]
So understanding it a bit better I can now say: it's clear to me that we it would be good to do something in this area. 
Either support mocking clients directly or introduce `RestMock`. 

Is tying it with `rest-client-reactive` much easier... What we need is the data from the resteasy reactive endpoint parser, no? Is it difficult to pull out?
The advantages of mocking clients are that it should be similarily simple (and in general, simple :) ) to integrate it with both rest clients, and an already known API (e.g. `Mockito`) could be used by the users. Obvious drawback is that we don't cover mocking APIs for which we don't use a JAX-RS based rest client.

**Comment by geoand** [comment]
> Is tying it with `rest-client-reactive` much easier... What we need is the data from the resteasy reactive endpoint parser, no? Is it difficult to pull out?

The issue is that the for resteasy-reactive this data is parsed and we can just consume it. 
With classic, the metadata isn't available anywhere so I would need to parse the interface all over again (with whatever bugs that might introduce)

**Comment by geoand** [comment]
After thinking about this more, I think we should try and introduce a dev-services type of flavor to this, meaning that it would be cool to make this work in dev as well as test mode.
It might complicate things, but I think it's worth it.

WDYT?

**Comment by geoand** [comment]
@FroMage @michalszynkiewicz @stuartwdouglas WDYT about ^?

**Comment by michalszynkiewicz** [comment]
It seems to me that, if we don't tie RestMocker to tests, a user can "simply" do sth like:
```java
@IfBuildProfile("dev")
class RestMockerInitializer {
   void onStart(@Observes StartupEvent ev) {
        RestMocker.when(service.getByName("GR")).replyWith(set);
        RestMocker.forward().toOriginalHost();
   }
}
```

@geoand do you mean sth like this?

**Comment by geoand** [comment]
Sure, but why would we want to put a user throught that?

My point is basically that the API Stef came up with is very lean and easy to use, to I am thinking that I would like to leverage it in dev-mode as well.

**Comment by michalszynkiewicz** [comment]
BTW, I realized that `RestMocker` has one more benefit over mocking the client, one can easily test headers that are sent.
MP Rest Client allows setting headers this way:
```
    @POST
    @ClientHeaderParam(name="X-Request-ID", value="{generateRequestId}")
    Response postWithRequestId(MyEntity entity);
```
With `RestMocker` we can provide means to easily verify the header value is as expected, with mocking the client with Mockito, we cannot.

**Comment by michalszynkiewicz** [comment]
By something like this, I didn't mean precisely this, but somethig that would do this under the hood. 
My thinking is that a user has to put their code somewhere. To use it in dev mode, it has to be production code, right?

**Comment by geoand** [comment]
Ah, you mean that the use of `@RestMock` would be in production instead of test code?

**Comment by michalszynkiewicz** [comment]
If we wanted to have it for dev mode too, it seems to be the best way.
How do you imagine it with "dev-services type of flavor"?

**Comment by geoand** [comment]
Yeah, that's what I am thinking of. It's not dev-services per-se as it's not zero-config, but the spirit of what I am thinking is indeed "dev-services type of flavor"

**Comment by FroMage** [comment]
This would indeed be very useful for dev mode, but I wonder how we can define what remote services we can mock, and what sort of responses would be OK.

IIRC @stuartwdouglas had the idea that dev mode for a micro-service would automatically fire up all other MS somehow. But this would work for MS the application would have code for, and I'm really not sure how it'd work. For MS like github, where we hit it but don't own it, we can't spawn it for real. We could say it's enough to hit it, after all, if there's network.

Perhaps we can do like Mockito and return 200 OK with empty (or default) responses for all mocked requests by default unless specified otherwise? That's the equivalent of the default return values for Mockito mocks…

**Comment by geoand** [comment]
> Perhaps we can do like Mockito and return 200 OK with empty (or default) responses for all mocked requests by default unless specified otherwise? That's the equivalent of the default return values for Mockito mocks…

+1

**Comment by geoand** [comment]
> IIRC @stuartwdouglas had the idea that dev mode for a micro-service would automatically fire up all other MS somehow. But this would work for MS the application would have code for, and I'm really not sure how it'd work. For MS like github, where we hit it but don't own it, we can't spawn it for real. We could say it's enough to hit it, after all, if there's network.
> 

Yeah, I'm thinking to do this only for `@RestMock` targets

> Perhaps we can do like Mockito and return 200 OK with empty (or default) responses for all mocked requests by default unless specified otherwise? That's the equivalent of the default return values for Mockito mocks…

+1

**Comment by michalszynkiewicz** [comment]
So if a user has an app with unconfigured URL (or mistyped a property name for URL), we will start a mock service that would return empty response for everything?
I'm not sure if that wouldn't be confusing for users, if they start getting weird errors (from empty responses) when they misconfigure the external service in dev mode...

I really like the idea of mocking for dev mode but IMO it should rather be opt-in.
And I would really like the interface for configuring the mocks to be programmatic, as Stef proposed with `RestMocker` and not some json/yaml

**Comment by geoand** [comment]
> So if a user has an app with unconfigured URL (or mistyped a property name for URL), we will start a mock service that would return empty response for everything?
> I'm not sure if that wouldn't be confusing for users, if they start getting weird errors (from empty responses) when they misconfigure the external service in dev mode...
> 
> I really like the idea of mocking for dev mode but IMO it should rather be opt-in.

I completely agree, hence the `@RestMock` API :)

**Comment by michalszynkiewicz** [comment]
After the initial scepticism, having more understanding of the use cases, I like the idea more and more :)

**Comment by geoand** [comment]
@stuartwdouglas any thoughts on the topic?

**Comment by stuartwdouglas** [comment]
I am not super convinced. I think my main concern here is that if you are mocking with the same classes you are using in the implementation then any errors might simply 'cancel out' and tests will pass when they shouldn't. For example in the 'Country' class above if I have misspelled 'capital` as 'capitol' then this won't be picked up by the tests, as both the client and the mock are using the same data structures. 

I have had other similar ideas about testing in the past, but have never taken them any further, for example if you have HelloResource:

```
@Path("/hello")
public class HelloResource {

    @GET
    @Produces(MediaType.TEXT_PLAIN)
    public String hello() {
        return "hello";
    }
}
```

I have thought that we could auto generate a REST client based on the resource:
```
@Inject
@TestRESTClient
HelloResource resource;
```

And then you could just use this directly to invoke the real HelloResource. The problem here is that any mistake in HelloResource would be replicated to the client, so there is a whole class of errors that you would miss with this approach.

**Comment by geoand** [comment]
Yeah, that is indeed a serious issue... Not sure how we could overcome that... I suppose that is the point of contract testing.
I just can't shake the feeling the there is a good solution out there that we just haven't stumbled upon yet...

**Comment by FroMage** [comment]
I don't think this is meant to test services for which we already have an implementation, but rather external services, for whom we could generate the rest client interface if we had the openapi definition, for example?

**Comment by geoand** [comment]
This is kinda confusuing me as to what it should and should not test and how we would explain it to users

**Comment by FroMage** [comment]
Well, look at this issue's original description, taken from our guide: we're mocking an external service. Not one we have access to any implementation for. We have a PR under way where we can generate a REST client interface from an openapi descriptor. When we have this, we can generate it and mock it, and in theory we couldn't get the model wrong.

**Comment by geoand** [comment]
Then perhaps we should revisit this when that is in?

**Comment by FroMage** [comment]
Sure. It's #16401.

**Comment by ssaip** [comment]
Any update on how a MP Rest Client can be mocked, now that the quarkus-openapi-generator is in place?

**Comment by geoand** [comment]
Reading this again after all this time, I keep coming back to:

> with testing what reaches a rest client, wouldn't someone just mock the client?

I can't shake the feeling that it doesn't make a lot of sense to try and solve this in the ways we have mentioned

**Comment by ssaip** [comment]
> with testing what reaches a rest client, wouldn't someone just mock the client?

agreed.


Originally I wanted to mock a response of the rest client, but also realized it doesn't make any sense to mock the client, so instead used WireMock to achieve what I wanted. This way I can also control response codes, headers, etc. (as @FroMage also already pointed out)

We already have this capability with WireMock.

Going through the original post again, I think (feel free to correct me) the suggestion was to only improve the UX of using WireMock or creating a "RestMocker" with an easy-to-use DSL and creating/returning actual objects instead of hard-to-read/maintain JSON.

As for the **"dev-services type of flavor"**: 
I think that's also a very needed feature for both test and dev in microservice environments and could probably utilize the same mechanisms.

My 2c on that:
> So if a user has an app with unconfigured URL (or mistyped a property name for URL), we will start a mock service that would return empty response for everything?

I'm against that, but -

> Perhaps we can do like Mockito and return 200 OK with empty (or default) responses for all mocked requests by default unless specified otherwise? That's the equivalent of the default return values for Mockito mocks…

I could imagine something like a reserved hostname (`http://mocked`) for that.
Might be worth to move this discussion into its own issue, though?

**Comment by geoand** [comment]
> Might be worth to move this discussion into its own issue, though?

Yeah, I think that makes sense
